src/TestSuite/Constraint/Email/MailConstraintBase.php

```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.7.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\TestSuite\Constraint\Email;
+
+use Cake\TestSuite\TestEmailTransport;
+use PHPUnit\Framework\Constraint\Constraint;
+
+/**
+ * Base class for all mail assertion constraints
```

Comment:
Should we mark these new constraint classes as `@internal` so they don't become part of our public API?

Reply:
I was going to bring that up that question as I'm going through the test suite and doing some work on it -- part of which is converting the assert methods in `(Console)IntegrationTestCase` to proper constraints.

EventFired and EventFiredWith are currently public; do we have to wait until 4.0 to mark those as `@internal`? The odds of them being used directly is likely slim.

Reply:
> EventFired and EventFiredWith are currently public; do we have to wait until 4.0 to mark those as @internal

No, we can do it now. We forgot earlier.